Stack/Opc.Ua.Core/Stack/Tcp/TcpTransportListener.cs

```diff
@@ -37,6 +38,233 @@ public override ITransportListener Create()
+            {
+                return m_blockedUntilTicks;
+            }
+            set
+            {
+                m_blockedUntilTicks = value;
+            }
+        }
+        #endregion
+
+        #region Private members
+        int m_lastRogueActionTicks;
+        int m_rogueActionCount;
+        int m_blockedUntilTicks;
+        #endregion
+    }
+
+    /// <summary>
+    /// Manages potential rogue clients
+    /// </summary>
+    public class RogueClientTracker : IDisposable
+    {
+        #region Public
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public RogueClientTracker()
+        {
+            m_cleanupTimer = new Timer(CleanupExpiredEntries, null, m_kCleanupIntervalMs, m_kCleanupIntervalMs);
+        }
+
+        /// <summary>
+        /// Checks if an IP address is currently blocked
+        /// </summary>
+        /// <param name="ipAddress"></param>
+        /// <returns></returns>
+        public bool IsBlocked(IPAddress ipAddress)
+        {
+            if (m_rogueClients.TryGetValue(ipAddress, out RogueClient client))
+            {
+                int currentTicks = HiResClock.TickCount;
+                return IsBlockedTicks(client.BlockedUntilTicks, currentTicks);
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Adds an action entry for a potential rogue client
+        /// </summary>
+        /// <param name="ipAddress"></param>
+        public void AddRogueClientAction(IPAddress ipAddress)
+        {
+            int currentTicks = HiResClock.TickCount;
+
+            m_rogueClients.AddOrUpdate(ipAddress,
+                // If client is new , create a new entry
+                key => new RogueClient {
+                    LastRogueActionTicks = currentTicks,
+                    RogueActionCount = 1,
+                    BlockedUntilTicks = 0
+                },
+                // If the client exists, update its entry
+                (key, existingEntry) => {
+                    // If IP currently blocked simply do nothing
+                    if (IsBlockedTicks(existingEntry.BlockedUntilTicks, currentTicks))
+                    {
+                        return existingEntry;
+                    }
+
+                    // Elapsed time since last recorded action
+                    int elapsedSinceLastRecAction = currentTicks - existingEntry.LastRogueActionTicks;
+
+                    if (elapsedSinceLastRecAction <= m_kRogueIntervalMs)
+                    {
+                        existingEntry.RogueActionCount++;
+
+                        if (existingEntry.RogueActionCount > m_kNrActionsTillBlock)
+                        {
+                            // Block the IP
+                            existingEntry.BlockedUntilTicks = currentTicks + m_kBlockDurationMs;
+                            Utils.LogError("RemoteClient IPAddress: {0} blocked for {1} ms due to exceeding {2} rogue actions under {3} ms ",
```

Comment:
due to {2} failed connection attempts within {3} ms
```suggestion
                            Utils.LogError("RemoteClient IPAddress: {0} blocked for {1} ms due to exceeding {2} rogue actions under {3} ms ",
```